ton_client/src/net/server_link.rs

```diff
@@ -407,31 +439,44 @@ impl ServerLink {
             headers.insert(name, value);
         }
 
-        let result = self
-            .client_env
-            .fetch(
-                &endpoint.query_url,
-                FetchMethod::Post,
-                Some(headers),
-                Some(request),
-                timeout,
-            )
-            .await;
+        let network_retries_count = self.config.network_retries_count;
+        let mut retry_count = 0;
+        'retries: loop {
+            let endpoint = if let Some(endpoint) = endpoint {
```

Comment:
Don't clone endpoint each time. Write this outside of loop 
`let query_url = endpoint.map(|endpoint| &endpoint.query_url)`.
Inside loop change this condition to this
```
if let Some(query_url) = query_url {
                query_url
            } else {
                &self.state.get_query_endpoint().await?.query_url
            };
```